decision/neighbour_filtering.py

```diff
@@ -40,24 +43,54 @@ def filtering_neighbours(state, number_of_neighbours=3, random_neighbours=0):
     total_score_list = []
     for cluster in clusters:
         cluster_id = cluster.id
-        if cluster_id != state.current_cluster.id:
+        if cluster_id != state.current_location.id:
             total_score = distance_scores[cluster_id] + deviation_scores[cluster_id]
             index = bisect.bisect(total_score_list, total_score)
             total_score_list.insert(index, total_score)
             score_indices.insert(index, cluster_id)
 
-    return (
-        [
+    if random_neighbours > 0:
+        neighbours = [
             clusters[index]
             for index in score_indices[: number_of_neighbours - random_neighbours]
-        ]
-        + [
+        ] + [
             clusters[
                 np.random.choice(
                     score_indices[number_of_neighbours - random_neighbours :]
                 )
             ]
```

Comment:
Do we only add one random neighbor?

Reply:
We add the number of random neighbours as the input parameter number_of_random_neighbours tells. The total number of neighbours is equal to number_of_neighbours, so if number_of_neighbours=5 and number_of_random_neighbours=2 -> return 3 "best" neighbours and 2 random neighbours

Reply:
I agree, but how can `clusters[np.random.choice(score_indices[number_of_neighbours - random_neighbours :])]` return more than one item? Check for yourself running `np.random.choice([1,2,4,5][:2])` in your console, it will only return 1 or 2. Maybe use the python random function `random.choices(clusters, k=(number_of_neighbours - random_neighbours))`. Maybe check that you choose from a list not containing the already chosen best neighbours?